KEYCLOAK-14202 Client Policy - Executor : Enforce secure signature algorithm for Signed JWT client authentication

tnorimat opened the pull request:
This PR is for [KEYCLOAK-14202 Client Policy - Executor : Enforce secure signature algorithm for Signed JWT client authentication](https://issues.redhat.com/browse/KEYCLOAK-14202) in [KEYCLOAK-13933 Client Policies](https://issues.redhat.com/browse/KEYCLOAK-13933), also is the part of the project [Client Policy Official Support](https://github.com/keycloak/kc-sig-fapi/projects) of [FAPI-SIG](https://github.com/keycloak/kc-sig-fapi) activity.

Generally speaking, the aim of this PR is to support policy conditions defined in [Client Policy design document](https://github.com/keycloak/keycloak-community/blob/master/design/client-policies.md#condition--which-client)

It behavior has been specified in [the JIRA ticket](https://issues.redhat.com/browse/KEYCLOAK-14202).

This PR has been written with the contribution by @andriimurashkin as FAPI-SIG activities..

tnorimat committed:
KEYCLOAK-14202 Client Policy - Executor : Enforce secure signature algorithm for Signed JWT client authentication

Co-authored-by: Andrii Murashkin <amu@adorsys.com.ua>

tnorimat commented:
@mposolda This PR is about Client Policy Official Support I mentioned in [page 16 of FAPI-SIG 6th Meeting Agenda](https://github.com/keycloak/kc-sig-fapi/blob/master/FAPI-SIG/meetings/6th/presentations/FAPI-SIG_6th_MTG_agenda%20.pdf). Could you or your team review this PR?

mposolda commented on services/src/main/java/org/keycloak/services/clientpolicy/executor/SecureSigningAlgorithmForSignedJwtEnforceExecutorFactory.java:
```diff
@@ -0,0 +1,65 @@
+
+    @Override
+    public String getHelpText() {
+        return "It prohibits the client whose JWT token signature algorithms are considered not to be secure.";
```
I know it has some side-effects and risk of outdated to enum the algorithms here, but still I would probably do it? Typical administrator, who will try to use this executor, may not know what the "signature algorithm considered to be secured" means. My vote is to add just another sentence like "Secured algorithms are PS and ES algorithms like for example PS256 or ES256." WDYT?

tnorimat replied:
Yes. I would like to describe the algorithms that are thought to be secure at this time explicitly, the same as `SecureSigningAlgorithmEnforceExecutorFactory.getHelpText()`

mposolda commented on services/src/main/java/org/keycloak/services/clientpolicy/executor/SecureSigningAlgorithmForSignedJwtEnforceExecutor.java:
```diff
@@ -0,0 +1,97 @@
+            return;
+        }
+        switch (signatureAlgorithm) {
+            case Algorithm.PS256:
```
I would add comment to source code like "// Please change also SecureSigningAlgorithmForSignedJwtEnforceExecutorFactory.getHelpText() if you are changing any algorithms here" .

See my other comment for the details.

tnorimat replied:
Yes, I agree with you. I'll fix as you pointed out.

mposolda reviewed (changes requested):
@tnorimat Thanks! Added a two minor comments inline. WDYT?

tnorimat commented:
@mposolda I've fixed codes and rebased.
As an aside of it, IMO, there are following tasks to complete the implementation of conditions and executors for FAPI-RW security profile support (to say more precisely, the last three are optional ones).

* [KEYCLOAK-14200 Client Policy - Executor : Enforce Holder-of-Key Token](https://issues.redhat.com/browse/KEYCLOAK-14200)
* [KEYCLOAK-16249 Client Policy - Condition : Client - Any Client](https://issues.redhat.com/browse/KEYCLOAK-16249)
* (no jira ticket yet) fix minor defects of all already merged conditions and executors for FAPI-RW
* (no jira ticket yet) revise test codes (ClientPolicyBasicsTest) because it included a lot.
* (no jira ticket yet) negative logic of the condition

As FAPI-SIG activities, I would like to try them one by one afterwards. It would be happy if you review them as usual. I think the last one is not so obvious task.

mposolda reviewed (approved):
@tnorimat Thanks!
